azure-core/azure/core/settings.py

```diff
@@ -0,0 +1,273 @@
+        if self._default is not _Unset:
+            return self._convert(self._default)
+
+        raise RuntimeError("No configured value found for setting %r" % self._name)
+
+    def __get__(self, instance, owner):
+        return self
+
+    def __set__(self, instance, value):
+        self.set_value(value)
+
+    def set_value(self, value):
+        # (Any) -> None
+        """Specify a value for this setting programmatically.
+
+        A value set this way takes precedence over all other methods except
+        immediate values.
+
+        :param time: a user-set value for this setting
+        :type time: str or int or float
+        :returns: None
+
+        """
+        self._user_value = value
+
+
+class Settings(object):
+    """Settings for globally used Azure configuration values.
+
+    The following methods are searched in order for a setting:
+
+    4. immediate values
+    3. previously user-set value
+    2. environment variable
+    1. system setting
+    0. implicit default
+
+    An implicit default is (optionally) defined by the setting attribute itself.
+
+    A system setting value can be obtained from registries or other OS configuration
+    for settings that support that method.
+
+    An environment variable value is obtained from ``os.environ``
+
+    User-set values many be specified by assigning to the attribute:
+
+    .. code-block:: python
+
+        settings.log_level = log_level = logging.DEBUG
+
+    Immediate values are (optionally) provided when the setting is retrieved:
+
+    .. code-block:: python
+
+        settings.log_level(logging.DEBUG()
+
+    Immediate values are most often useful to provide from optional arguments
+    to client functions. If the argument value is not None, it will be returned
+    as-is. Otherwise, the setting searches other methods according to the
+    precedence rules.
+
+    :Attributes:
+
+    :cvar log_level: a log level to use across all Azure client SDKs (AZURE_LOG_LEVEL)
+    :type log_level: PrioritizedSetting
+    :cvar proxy_settings: proxy settings to use across all Axure client SDKs (AZURE_PROXY_SETTINGS)
+    :type proxy_settings: PrioritizedSetting
+    :cvar telemetry_enabled: Whether telemtry should run across Azure SDKs (AZURE_TELEMETRY_ENABLED)
+    :type telemetry_enabled: PrioritizedSetting
+    :cvar tracing_enabled: Whether tracing shoudl be enabled across Azure SDKs (AZURE_TRACING_ENABLED)
+    :type tracing_enabled: PrioritizedSetting
+
+    :Example:
+
+    >>> import logging
+    >>> from azure.core.settings import settings
+    >>> settings.log_level = logging.DEBUG
+    >>> settings.log_level()
+    10
+
+    >>> settings.log_level(logging.WARN)
+    30
+
+    """
+
+    log_level = PrioritizedSetting(
+        "log_level",
+        env_var="AZURE_LOG_LEVEL",
+        convert=convert_logging,
+        default=logging.INFO,
+    )
+
+    proxy_settings = PrioritizedSetting(
+        "proxy_settings", env_var="AZURE_PROXY_SETTINGS"
```

Comment:
We should pick up the `HTTP`/`HTTPS_PROXY` environment variables for proxy settings. And `HTTP_NO_PROXY` environment variable.

Reply:
`HTTP_NO_PROXY` is a bool? perhaps `HTTP_PROXY_ENABLED` to match the sense and verbiage of the others, unless there's a precedent to match?

Reply:
The first part is fixed in https://github.com/Azure/azure-sdk-for-python/pull/5192/commits/099f2f9839f065c7dfd4bb4a6c0aff7c8067a6eb will address "no proxy" after clarification